### DISCUSSION TITLE
Type issue where field is not null

### DISCUSSION THREAD

**Discussion by MarkForLoop** [prisma/prisma#20190] [Q&A]
I am trying to access the someID property of the items object where someID is not null 

However, the type of the someID property is not being inferred correctly, and I have to filter out the null values from the resulting array.

Here is the code snippet that I am using:
```
const result = await ctx.prisma.example.findUnique({
  where: {
    id: input.id,
  },
  include: {
    items: {
      where: {
        someID: { not: null },
      },
      select: { someID: true },
    },
  },
});

if (result.items.length > 0) {
  const someIDs = result.items.map((item) => item.someID);
}
```

As you can see, the someIDs array is inferred as (string | null)[] instead of string[]. 
This is because the someID property can be null, and TypeScript infers the type of the resulting array as a union of string and null.

1.  is this intended? 
2. If not how do I fix it?

**Comment by nurul3101** [comment]
Hello @MarkForLoop 👋 

Can you please share your schema file?

If I understand correctly, if `someID` is defined as optional in the schema file, then this behaviour is intended because TypeScript is correctly inferring that the someID property could potentially be null.

**Comment by MarkForLoop** [comment]
This is the schema I am working it
```
model Example {
  id          String   @id @unique @default(uuid())
  title       String
  description String
  createdAt   DateTime @default(now())
  updatedAt   DateTime @default(now()) @updatedAt
  item        Item[]
}

model Item {
  id        String   @id @unique @default(uuid())
  title     String
  text      String
  exampleId String
  someID    String?
  createdAt DateTime @default(now())
  updatedAt DateTime @default(now()) @updatedAt
  example   Example  @relation(fields: [exampleId], references: [id])
  some      Some?     @relation(fields: [someID], references: [id])
}

model Some {
  id        String   @id @unique @default(uuid())
  type      MediaType
  url       String
  createdAt DateTime @default(now())
  updatedAt DateTime @default(now()) @updatedAt
  item      Item[]
}

enum MediaType {
  image
  video
  audio
}

```




I assumed that in the where clause since I filtered out  
```
where: { someID: { not: null } }
```
it would figure out that the `someID` is now a `string`, instead of being a `string | null`.

**Reply by nurul3101** [reply]
Thanks for sharing the schema file, I believe this is expected because `someID` is defined as optional.

It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 
Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by nurul3101** [reply]
We are closing this Discussion due to inactivity. Please open a new Discussion if you have a similar problem 🙏

**Reply by mr-scrpt** [reply]
What's normal about it, if the condition says to select values if the field is not null, then how can the result type tell us that the field can be null? It's nonsense

**Comment by mamlzy** [comment]
facing the same issue, typescript not infer the type, even using `not: null`

**Comment by patsissons** [comment]
came up with a quick type construct solution to this today

```ts
export type NonNullFields<T, K extends keyof T> = Omit<T, K> & {
  [K in keyof T]-?: NonNullable<T[K]>;
};

export type NonNullFieldsArray<T extends unknown[], K extends keyof T[number]> =
  T extends Array<infer U> ? NonNullFields<U, K>[] : never;
```

while not perfect, it at least reduces the problem down to 

```ts
prisma.thing
  .findMany({...})
  .then((things) => things as NonNullFieldsArray<typeof things, "non_null_field">)
```

**Reply by wtachau** [reply]
I think this needs to be 
```ts
export type NonNullFields<T, K extends keyof T> = Omit<T, K> & {
    [P in K]: NonNullable<T[P]>;
};
```
so that other fields of T remain nullable.

**Comment by metehandemir** [comment]
facing same issue.

**Comment by oxidia** [comment]
I am facing the same issue.

**Comment by afonsomatos** [comment]
Same

**Comment by Mnigos** [comment]
same

**Comment by Vimbi** [comment]
Reopen https://github.com/prisma/prisma/issues/26301

**Reply by ludralph** [reply]
Hi @Vimbi 

We will take a look at the issue you reopened and provide a response.

**Comment by chaples55** [comment]
also same

**Comment by castrosuellenx** [comment]
same

**Comment by Magnuti** [comment]
This is still highly relevant in 2025. It would be great if Prisma returned the correct type.

Here is an open issue discussing the problem https://github.com/prisma/prisma/issues/20348

**Comment by otakutyrant** [comment]
We can use `as` syntax to narrow the type manually.

**Comment by MaksimKonyshev** [comment]
It would be great to have something like that:

```
const result = await ctx.prisma.example.findMany({
  where: {
    isDeleted: false,
  },
  whereNonNullable: {
    someID: true,
  },
});
```

typeof example is `{ isDeleted: boolean, someID: number | null }`
but typeof result is `{ isDeleted: boolean, someID: number }`
